File: C++/binarytree.cpp

class BinaryTree {

  struct Tree {
    int value = 0;
    Tree* left = 0;
    Tree* right = 0;

    bool right_is_empty();
    bool left_is_empty();
  };

  Tree* root = new Tree;

  void insert(int value, Tree* root);
  bool contains(int value, Tree* root);

  void pacman(Tree* root);

public:
  BinaryTree(int rootValue);
  ~BinaryTree();

  void insert(int value);
  bool contains(int value);

};

BinaryTree::BinaryTree(int rootValue) {
  root->value = rootValue;
}

BinaryTree::~BinaryTree() {
  BinaryTree::pacman(root);
}

bool BinaryTree::Tree::left_is_empty() {
  return left == 0;
}

bool BinaryTree::Tree::right_is_empty() {
  return right == 0;
}

void BinaryTree::pacman(Tree* root) {
  if (!root->left_is_empty()) {
    pacman(root->left);
  }
  if (!root->right_is_empty()) {
    pacman(root->right);
  }
  delete root;
}

bool BinaryTree::contains(int value) {
  return contains(value, root);
}

bool BinaryTree::contains(int value, Tree* root) {
  if (value == root->value) {
    return true;
  }
  if (value < root->value && !root->left_is_empty()) {
    return contains(value, root->left);
  }
  if (value > root->value && !root->right_is_empty()) {
    return contains(value, root->right);
  }
  return false;
}

void BinaryTree::insert(int value) {
  insert(value, root);
}

void BinaryTree::insert(int value, Tree* root) {
  if (value > root->value) {
    if (root->right_is_empty()) {
      Tree* tmp = new Tree;
      tmp->value = value;
      root->right = tmp;
    }
    else {
      insert(value, root->right);
    }
  }
  else {
    if (root->left_is_empty()) {
      Tree* tmp = new Tree;
      tmp->value = value;
      root->left = tmp;
    }
    else {
      insert(value, root->left);
    }
  }
}

File: Go/binarytree.go
// This is a simple package for Golang that provides a very
// rudimentary implementation of a Binary Tree, provides insertion,
// deletion, and a search. To try to support some custom types instead
// of just limiting the usage to ints or strings, I've addded an
// interface that asks you to implement three comparison functions to
// enable the tree comparison steps. This could probably be done
// better but I'm just trying to get back into the swing of my
// favourite data structures.ma
package BinaryTree

import (
	"fmt"
)

type BTreeCompare interface {
	LessThan(other BTreeCompare) bool
	GreaterThan(other BTreeCompare) bool
	EqualTo(other BTreeCompare) bool
}

type BinaryTree struct {
	value BTreeCompare // value of this node
	left  *BinaryTree  // next smaller value
	right *BinaryTree  // next larger value
}

func (tree *BinaryTree) resetLeaf(value BTreeCompare, newChild *BinaryTree) {
	if value.GreaterThan(tree.value) {
		tree.right = newChild
	} else {
		tree.left = newChild
	}
}

func New(value BTreeCompare) *BinaryTree {
	t := new(BinaryTree)
	t.value = value
	return t
}

func (root *BinaryTree) Preorder() {
	preorder_print(root)
}

func preorder_print(root *BinaryTree) {
	if root != nil {
		fmt.Println(root.value)
		preorder(root.left)
		preorder(root.right)
	}
}

func (root *BinaryTree) Contains(value BTreeCompare) bool {
	switch {
	case value.EqualTo(root.value):
		return true
	case value.LessThan(root.value) && root.left != nil:
		return root.left.Contains(value)
	case value.GreaterThan(root.value) && root.right != nil:
		return root.right.Contains(value)
	}
	return false
}

func (root *BinaryTree) Delete(value BTreeCompare) {
	if root.Contains(value) {
		delete(value, root, nil)
	}
}

func (root *BinaryTree) Insert(value BTreeCompare) {
	switch {
	case value.GreaterThan(root.value) && root.right == nil:
		n := new(BinaryTree)
		n.value = value
		root.right = n
	case value.LessThan(root.value) && root.left == nil:
		n := new(BinaryTree)
		n.value = value
		root.left = n
	case value.GreaterThan(root.value):
		root.right.Insert(value)
	case value.LessThan(root.value):
		root.left.Insert(value)
	}
}

func delete(value BTreeCompare, current, previous *BinaryTree) {
	if value.EqualTo(current.value) {
		switch {
		case previous == nil && current.left == nil && current.right == nil:
			// Errmmmmmmm....
			current = new(BinaryTree)
		case current.left == nil && current.right == nil:
			previous.resetLeaf(value, nil)

		case current.left != nil:
			tmp := current.left
			tmp.right = current.right
			previous.resetLeaf(tmp.value, tmp)

		case current.right != nil:
			tmp := current.right
			previous.resetLeaf(current.right.value, current.right)
		}
	} else {
		switch {
		case value.LessThan(current.value) && current.left != nil:
			delete(value, current.left, current)
		case value.GreaterThan(current.value) && current.right != nil:
			delete(value, current.right, current)
		}
	}
}
